psydac/linalg/kernels.py

```diff
@@ -0,0 +1,163 @@
+                    data[ind] = value
+                    ind += 1
+    return ind
+
+def stencil2coo_3d_C(A:'float64[:,:,:,:,:,:]', data:'float64[:]', rows:'int64[:]', cols:'int64[:]',
+                     nrl1:'int64', nrl2:'int64', nrl3:'int64', ncl1:'int64', ncl2:'int64', ncl3:'int64',
+                     s1:'int64', s2:'int64', s3:'int64', nr1:'int64', nr2:'int64', nr3:'int64',
+                     nc1:'int64', nc2:'int64', nc3:'int64', dm1:'int64', dm2:'int64', dm3:'int64',
+                     cm1:'int64', cm2:'int64', cm3:'int64', p1:'int64', p2:'int64', p3:'int64',
+                     dp1:'int64', dp2:'int64', dp3:'int64'):
+    ind = 0
+    pp1 = cm1*p1
+    pp2 = cm2*p2
+    pp3 = cm3*p3
+    for i1 in range(nrl1):
+        for i2 in range(nrl2):
+            for i3 in range(nrl3):
+                ii1 = s1+i1
+                ii2 = s2+i2
+                ii3 = s3+i3
+                I   = ii1*nr2*nr3 + ii2*nr3 + ii3
+                for j1 in range(ncl1):
+                    for j2 in range(ncl2):
+                        for j3 in range(ncl3):
+                            value = A[i1+pp1,i2+pp2,i3+pp3,j1,j2,j3]
+                            if abs(value)<1e-15:continue
+                            jj1 = ((ii1//cm1)*dm1+j1-dp1)%nc1
+                            jj2 = ((ii2//cm2)*dm2+j2-dp2)%nc2
+                            jj3 = ((ii3//cm3)*dm3+j3-dp3)%nc3
+
+                            J   = jj1*nc2*nc3 + jj2*nc3 + jj3
+
+                            rows[ind] = I
+                            cols[ind] = J
+                            data[ind] = value
+                            ind += 1
+
+    return ind
+
+def stencil2coo_3d_F(A:'float64[:,:,:,:,:,:]', data:'float64[:]', rows:'int64[:]', cols:'int64[:]',
+                     nrl1:'int64', nrl2:'int64', nrl3:'int64', ncl1:'int64', ncl2:'int64', ncl3:'int64',
+                     s1:'int64', s2:'int64', s3:'int64', nr1:'int64', nr2:'int64', nr3:'int64',
+                     nc1:'int64', nc2:'int64', nc3:'int64', dm1:'int64', dm2:'int64', dm3:'int64',
+                     cm1:'int64', cm2:'int64', cm3:'int64', p1:'int64', p2:'int64', p3:'int64',
+                     dp1:'int64', dp2:'int64', dp3:'int64'):
+    ind = 0
+    pp1 = cm1*p1
+    pp2 = cm2*p2
+    pp3 = cm3*p3
+    for i1 in range(nrl1):
+        for i2 in range(nrl2):
+            for i3 in range(nrl3):
+                ii1 = s1+i1
+                ii2 = s2+i2
+                ii3 = s3+i3
+                I   = ii3*nr1*nr2 + ii2*nr1 + ii1
+                for j1 in range(ncl1):
+                    for j2 in range(ncl2):
+                        for j3 in range(ncl3):
+                            value = A[i1+pp1,i2+pp2,i3+pp3,j1,j2,j3]
+                            if abs(value)<1e-15:continue
+                            jj1 = ((ii1//cm1)*dm1+j1-dp1)%nc1
+                            jj2 = ((ii2//cm2)*dm2+j2-dp2)%nc2
+                            jj3 = ((ii3//cm3)*dm3+j3-dp3)%nc3
```

Comment:
The modulo operator `%` is expensive, and we might want to avoid it when the boundary conditions are not periodic. Maybe put a TODO item in a comment?

Reply:
ah yes, this can be improved

Reply:
done [7c15659](https://github.com/pyccel/psydac/pull/248/commits/7c15659b079742fedad0ff09b0450822e12cf815)